src/macaron/slsa_analyzer/checks/check_result.py

```diff
@@ -24,25 +29,54 @@ class CheckResultType(str, Enum):
     UNKNOWN = "UNKNOWN"
 
 
-class CheckResult(TypedDict):
-    """This class stores the result of a check in a dictionary."""
+@dataclass
+class CheckInfo:
+    """This class identifies and describes a check."""
 
     check_id: str
     check_description: str
-    # The string representations of the slsa requirements and their
-    # corresponding slsa level.
-    slsa_requirements: list[str]
+    eval_reqs: list[ReqName]
+
+
+@dataclass
+class CheckResultData:
+    """This class stores the result of a check."""
+
     # If an element in the justification is a string,
     # it will be displayed as a string, if it is a mapping,
     # the value will be rendered as a hyperlink in the html report.
-    justification: list[str | dict[str, str]]
+    justification: Justification
     # human_readable_justification: str
     # result_values: dict[str, str | float | int] | list[dict[str, str | float | int]]
-    result_tables: list[DeclarativeBase | Expectation]
+    result_tables: ResultTables
     # recommendation: str
     result_type: CheckResultType
 
 
+@dataclass
```

Comment:
True, better for them to be immutable. Dataclasses can be defined as immutable with `@dataclass(frozen=True)`, so I think it is preferable to do that rather than use `NamedTuple`.

Reply:
Curious why you prefer `@dataclass(frozen=True)` over `NamedTuple`? One nice thing about `NamedTuple` is that it supports unpacking and the types **must** be specified. That seems particularly useful for `CheckInfo` and `CheckResultData`.

Reply:
I read this (https://peps.python.org/pep-0557/#why-not-just-use-typing-namedtuple) and it seemed that `dataclass` was the more strongly-typed option, since `NamedTuple` falls back to just being a tuple in various cases.
As far as I can tell, you do have to declare all the fields (with types) for a `dataclass(frozen=True)`, since you need to provide types to declare the field, and even if you try and define an explicit constructor to avoid that it won't let you assign self fields.

Reply:
Yes, looks like `dataclass` also requires the attribute to be typed (explicitly or trough assignment).  I don't think one is better than the other in terms of typing. See examples below, both allowing a `str` value being assigned to an `int` attribute: :disappointed:  

```bash
>>> class A(NamedTuple):
...     foo: int
>>> a = A(foo="foo")
>>> type(a.foo)
<class 'str'>
```
vs
```bash
>>> @dataclass(frozen=True)
... class B():
...     foo: int
>>> b = B(foo="foo")
>>> type(b.foo)
<class 'str'>
```
So at this point, I guess there is no big difference.

Reply:
Yes, the static types are not enforced at runtime, but that's true of all python static typing. Mypy gives a type error in both cases, so it's as type-safe as python can be.

Reply:
I have changed them to `@dataclass(frozen=True)`.